skills/repo.go

```diff
@@ -157,6 +168,191 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
+func (r *Repository) Slot(snapName, slotName string) *Slot {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.slots[snapName][slotName]
+}
+
+// AddSlot adds a new slot to the repository.
+// Adding a slot with invalid name returns an error.
+// Adding a slot that has the same name and snap name as another slot returns an error.
+func (r *Repository) AddSlot(slot *Slot) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject snaps with invalid names
+	if err := snap.ValidateName(slot.Snap); err != nil {
+		return err
+	}
+	// Reject skill with invalid names
+	if err := ValidateName(slot.Name); err != nil {
+		return err
+	}
+	// TODO: ensure that apps are correct
+	t := r.types[slot.Type]
+	if t == nil {
+		return fmt.Errorf("cannot add slot, skill type %q is not known", slot.Type)
+	}
+	if _, ok := r.slots[slot.Snap][slot.Name]; ok {
+		return fmt.Errorf("cannot add slot, snap %q already has slot %q", slot.Snap, slot.Name)
+	}
+	if r.slots[slot.Snap] == nil {
+		r.slots[slot.Snap] = make(map[string]*Slot)
+	}
+	r.slots[slot.Snap][slot.Name] = slot
+	return nil
+}
+
+// RemoveSlot removes a named slot from the given snap.
+// Removing a slot that doesn't exist returns an error.
+// Removing a slot that uses a skill returns an error.
+func (r *Repository) RemoveSlot(snapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such slot exists
+	slot := r.slots[snapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot remove slot %q from snap %q, no such slot", slotName, snapName)
+	}
+	// Ensure that the slot is not using any skills
+	if len(r.skillUsedBySlot[slot]) > 0 {
+		return fmt.Errorf("cannot remove slot %q from snap %q, it still uses granted skills", slotName, snapName)
+	}
+	delete(r.slots[snapName], slotName)
+	if len(r.slots[snapName]) == 0 {
+		delete(r.slots, snapName)
+	}
+	return nil
+}
+
+// Grant grants the named skill to the named slot of the given snap.
+// The skill and the slot must have the same type.
+func (r *Repository) Grant(skillSnapName, skillName, slotSnapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such skill exists
+	skill := r.skills[skillSnapName][skillName]
+	if skill == nil {
+		return fmt.Errorf("cannot grant skill %q from snap %q, no such skill", skillName, skillSnapName)
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot grant skill to slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that skill and slot are compatible
+	if slot.Type != skill.Type {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q, skill type %q doesn't match slot type %q",
+			skillName, skillSnapName, slotName, slotSnapName, skill.Type, slot.Type)
+	}
+	// Ensure that slot and skill are not connected yet
+	if r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q twice",
```

Comment:
This should probably not be an error. The exact semantics the function call requested are already in place, so there's no reason to blow it up I think. We should just return nil, and the call sites will be happy that the state they want is active.